mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocFinder.java

```diff
@@ -97,50 +96,61 @@ public long find(CRUDOperationContext ctx,
             }
             int size = cursor.count();
             long ret = size;
-            LOGGER.debug("Applying limits: {} - {}", from, to);
 
-            if (from != null) {
-                cursor.skip(from.intValue());
-                if (to != null && from > to) //if 'to' is not null but lesser than 'from', skip the entire list to return nothing 
-                {
-                    cursor.skip(size);
+            int nRetrieve=size;
+            List<DBObject> mongoResults=null;            
+            List<JsonDoc> jsonDocs=null;
+
+           LOGGER.debug("Applying limits: {} - {}", from, to);
+            long retrievalTime=0;
+
+            // f and t are from and to indexes, both inclusive
+            int f=from==null?0:from.intValue();
+            int t=to==null?size-1:to.intValue();
+            if(f<0)
+                f=0;
+            if(t<0)
+                nRetrieve=0;
+            if(t>=size)
+                t=size-1;
+            if(f>=size) 
+                nRetrieve=0;
```

Comment:
I think the nRetrieve=0 cases are all essentially the same condition, but reworded, so maybe they could be collapsed.

- If `t < 0`, then `t` will be `< f`, because `f` is clamped at min 0. So you can just clamp `f` at 0 at then you don't need `t < 0` in addition to `t < f`.
- If `f` is `>= size`, then `t` will always be `< f`, because `t` is max size-1, so you don't need `f >= size` in addition to `t < f`.

So `t < f` covers both cases given the boundaries.

Ultimately I think this is all saying (leaving out null handling for brevity): 

1. set f = from min 0
2. set t = to max size -1
3. set nRetrieve = t-f+1 min 0

Then deal with nRetrieve accordingly as you already do. To me this algorithm is simpler to understand at a glance.

What do you think?